### PR TITLE
Add queue command

### PR DESCRIPTION
**Pull request by OtterBoops**
Basic list of songs queued, including the one currently playing.
Stuff to be taken care of after initial approval:

- [ ] - Sanitize output
- [x] - Format better
- [x] - Pagination (somehow)
- [ ] - Add requester info

### CODE REVIEW

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        reply(printQueue(players[guild].getQueue()))
+    }
+
+    private fun CommandContext.printQueue(queue: List<AudioTrack>): String {
+        val sb = StringBuilder()
+        when (queue.size) {
+            0 -> sb.append("Empty queue")
+            else -> {
+                queue.forEachIndexed {
+                    index, t -> sb.append("`[${index+1}]` ${t.info.title} \n")
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:24]: `index, t ->` usually goes on the first line

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        private val players: PlayerRegistry
+) : Command("q") {
+
+    override suspend fun CommandContext.invoke() {
+        reply(printQueue(players[guild].getQueue()))
+    }
+
+    private fun CommandContext.printQueue(queue: List<AudioTrack>): String {
+        val sb = StringBuilder()
+        when (queue.size) {
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:20]: Having a when block doesn't make sense here.
1. You can just do `if (queue.isEmpty() return "Empty queue"`
2. Even then, an if/else would look nicer

**Review by Frederikam — CHANGES_REQUESTED** [review]

**Comment by OtterBoops** [comment]
Off the top of your head, can the current way of writing the list cause any issues (unwanted mentions, janky text, etc.)?

Also is there a way for me to get the song requester that is readily available? Or does that need to be implemented separately?

**Comment by Frederikam** [comment]
Sorry for the late reply. I'm currently sick.

> Off the top of your head, can the current way of writing the list cause any issues (unwanted mentions, janky text, etc.)?

All mentions are configured to be disabled by default. Janky text is possible. At some point I would like to write formatter helper functions that take unsafe args as varargs. Don't worry about it for now.

> Also is there a way for me to get the song requester that is readily available? Or does that need to be implemented separately?

We'd need to store metadata about it. Lavaplayer is in no way aware of JDA. Lavaplayer has a way to set custom data with each track. I have a fancy way of doing this with Kotlin.

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+         
+        return hms
+    }
+
+    private fun CommandContext.paginateQueue(queue: List<AudioTrack>, index: Int) {
+        val pageCount: Int = (queue.size + pageSize - 1) / pageSize
+        val pageIndex: Int 
+
+        // Handle edge cases for arguments
+        if (index == 0)
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:61]: This chain can be replaced by a single call to `Int#coerceIn()`

If you were using Java, I would recommend `Math.max()` and `Math.min()` on a single line.

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        // Handle edge cases for arguments
+        if (index == 0) 
+            pageIndex = 1
+        else if (index > pageCount) 
+            pageIndex = pageCount
+        else
+            pageIndex = index
+
+        //Add header
+        response.append("Page **${pageIndex}** of **${pageCount}**\n\n")
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:69]:
```suggestion
        response.append("Page **$pageIndex** of **$pageCount**\n\n")
```

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        } else {
+            hms = String.format("%02d:%02d:%02d", TimeUnit.MILLISECONDS.toHours(length),
+            TimeUnit.MILLISECONDS.toMinutes(length) % TimeUnit.HOURS.toMinutes(1),
+            TimeUnit.MILLISECONDS.toSeconds(length) % TimeUnit.MINUTES.toSeconds(1));
+        }
+         
+        return hms
+    }
+
+    private fun CommandContext.paginateQueue(queue: List<AudioTrack>, index: Int) {
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:56]: Try this neat helper function to use a StringBuilder as receiver and return a string
```suggestion
    private fun CommandContext.paginateQueue(queue: List<AudioTrack>, index: Int) = builder {
```

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            pageIndex = pageCount
+        else
+            pageIndex = index
+
+        //Add header
+        response.append("Page **${pageIndex}** of **${pageCount}**\n\n")
+
+        // Preserve original indexes
+        queue.forEachIndexed { ind, t ->
+            if (t in queue.chunked(pageSize)[pageIndex - 1])
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:73]: Wouldn't it be smarter to just iterate a sublist?

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            argumentText.toIntOrNull() != null -> reply(printQueue(queue, argumentText.toInt()))
+        }
+        response.clear()
+    }
+
+    private fun CommandContext.printQueue(queue: List<AudioTrack>, pageIndex: Int): String {
+        val totalDuration = players[guild].getDuration()
+        
+        if (queue.isEmpty())
+            return "Not currently playing anything."
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:33]:
```suggestion
            return "The queue is empty."
```

File: src/main/kotlin/dev/arbjerg/ukulele/audio/TrackQueue.kt [hunk 8]
```diff
@@ -17,5 +17,14 @@ class TrackQueue {
         queue.removeAll(list)
         return list
     }
-
-}
\ No newline at end of file
+    fun getQueue(): List<AudioTrack> {
+        return queue
+    }
+    fun getDuration(): Long {
```
Comment (Frederikam) [inline src/main/kotlin/dev/arbjerg/ukulele/audio/TrackQueue.kt:23]: Want to save 6 lines of code?
`fun getDuration() = queue.sum { it.info.length }`

**Review by Frederikam — CHANGES_REQUESTED** [review]

File: src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt (hunk 6 above)
Comment (OtterBoops) [reply src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt:73]: Wouldn't a sublist  hold its' own indexes, in this case 0-4? Or am I mistaken?

Comment (Frederikam) [reply src/main/kotlin/dev/arbjerg/ukulele/command/QueueCommand.kt]: Yep, but we can just calculate an offset

**Review by Frederikam — APPROVED** [review]

**Pull request merged**